File: server/services/progressService.ts
import { db } from "../db";
import { studyProgress } from "../../shared/schema";
import { eq, and } from "drizzle-orm";

export interface ProgressUpdate {
  userId: string;
  sessionId: string;
  topicsReviewed: string[];
  comprehensionLevel: number; // 0-100
  timeSpent: number; // in minutes
  questionsAnswered?: number;
  correctAnswers?: number;
}

export interface ProgressStats {
  totalSessions: number;
  totalTimeSpent: number;
  averageComprehension: number;
  topicsReviewed: string[];
  lastSessionDate: Date | null;
}

/**
 * Track user's study progress
 */
export async function trackProgress(
  userId: string,
  sessionId: string,
  data: Omit<ProgressUpdate, "userId" | "sessionId">
): Promise<void> {
  try {
    const topicsString = JSON.stringify(data.topicsReviewed);
    
    await db.insert(studyProgress).values({
      userId,
      sessionId,
      topicsReviewed: topicsString,
      comprehensionLevel: data.comprehensionLevel,
      timeSpent: data.timeSpent,
      questionsAnswered: data.questionsAnswered || 0,
      correctAnswers: data.correctAnswers || 0,
      createdAt: new Date(),
    });
  } catch (error) {
    console.error("Error tracking progress:", error);
    throw error;
  }
}

/**
 * Get user's progress statistics
 */
export async function getUserProgressStats(userId: string): Promise<ProgressStats> {
  try {
    const allProgress = await db
      .select()
      .from(studyProgress)
      .where(eq(studyProgress.userId, userId));

    if (allProgress.length === 0) {
      return {
        totalSessions: 0,
        totalTimeSpent: 0,
        averageComprehension: 0,
        topicsReviewed: [],
        lastSessionDate: null,
      };
    }

    const totalSessions = allProgress.length;
    const totalTimeSpent = allProgress.reduce((sum, p) => sum + (p.timeSpent ?? 0), 0);
    const averageComprehension = 
      allProgress.reduce((sum, p) => sum + (p.comprehensionLevel ?? 0), 0) / totalSessions;

    // Collect all unique topics
    const allTopics = new Set<string>();
    allProgress.forEach((p) => {
      try {
        if (p.topicsReviewed) {
          const topicsStr = typeof p.topicsReviewed === 'string' ? p.topicsReviewed : JSON.stringify(p.topicsReviewed);
          const topics = JSON.parse(topicsStr);
          topics.forEach((t: string) => allTopics.add(t));
        }
      } catch {
        // Skip if JSON parsing fails
      }
    });

    // Get last session date
    const lastSessionDate = allProgress.reduce((latest, current) => {
      const currentDate = current.createdAt ? new Date(current.createdAt) : new Date();
      return currentDate > latest ? currentDate : latest;
    }, allProgress[0].createdAt ? new Date(allProgress[0].createdAt) : new Date());

    return {
      totalSessions,
      totalTimeSpent,
      averageComprehension: Math.round(averageComprehension * 100) / 100,
      topicsReviewed: Array.from(allTopics),
      lastSessionDate,
    };
  } catch (error) {
    console.error("Error getting progress stats:", error);
    throw error;
  }
}

/**
 * Get session-specific progress
 */
export async function getSessionProgress(sessionId: string) {
  try {
    const progress = await db
      .select()
      .from(studyProgress)
      .where(eq(studyProgress.sessionId, sessionId));

    return progress.length > 0 ? progress[0] : null;
  } catch (error) {
    console.error("Error getting session progress:", error);
    throw error;
  }
}

/**
 * Update comprehension level for a session
 */
export async function updateComprehensionLevel(
  sessionId: string,
  comprehensionLevel: number
): Promise<void> {
  try {
    if (comprehensionLevel < 0 || comprehensionLevel > 100) {
      throw new Error("Comprehension level must be between 0 and 100");
    }

    await db
      .update(studyProgress)
      .set({
        comprehensionLevel,
        updatedAt: new Date(),
      })
      .where(eq(studyProgress.sessionId, sessionId));
  } catch (error) {
    console.error("Error updating comprehension level:", error);
    throw error;
  }
}

File: services/progress_service.py
from typing import Dict, List, Optional
from datetime import datetime
from dataclasses import dataclass


@dataclass
class ProgressStats:
    total_sessions: int
    total_time_spent: int
    average_comprehension: float
    topics_reviewed: List[str]
    last_session_date: Optional[str]


def format_progress_stats(stats: ProgressStats) -> Dict:
    """Convert progress stats to dictionary"""
    return {
        'total_sessions': stats.total_sessions,
        'total_time_spent': stats.total_time_spent,
        'average_comprehension': stats.average_comprehension,
        'topics_reviewed': stats.topics_reviewed,
        'last_session_date': stats.last_session_date
    }


async def track_progress(
    user_id: str,
    session_id: str,
    topics_reviewed: List[str],
    comprehension_level: int,
    time_spent: int,
    questions_answered: int = 0,
    correct_answers: int = 0
) -> bool:
    print(f"Recording progress for user {user_id}, session {session_id}")
    print(f"  Topics: {topics_reviewed}")
    print(f"  Comprehension: {comprehension_level}%")
    print(f"  Time: {time_spent} minutes")
    
    return True


async def get_user_progress_stats(user_id: str) -> Optional[ProgressStats]:
    """Get user progress statistics"""
    return ProgressStats(
        total_sessions=0,
        total_time_spent=0,
        average_comprehension=0.0,
        topics_reviewed=[],
        last_session_date=None
    )


async def get_session_progress(session_id: str) -> Optional[Dict]:
    print(f"Fetching progress for session {session_id}")
    # This will be implemented with database connection
    return None


async def update_comprehension_level(
    session_id: str,
    comprehension_level: int
) -> bool:
    if comprehension_level < 0 or comprehension_level > 100:
        raise ValueError("Comprehension level must be between 0 and 100")
    
    print(f"Updating comprehension for session {session_id}: {comprehension_level}%")
    # This will be implemented with database connection
    return True


def calculate_average_comprehension(comprehension_levels: List[int]) -> float:
    if not comprehension_levels:
        return 0.0
    return sum(comprehension_levels) / len(comprehension_levels)


def aggregate_topics(all_topics_lists: List[List[str]]) -> List[str]:
    unique_topics = set()
    for topics in all_topics_lists:
        unique_topics.update(topics)
    return sorted(list(unique_topics))
